core/consensus/babe/src/lib.rs

```diff
@@ -537,20 +537,45 @@ fn check_header<B: BlockT + Sized, C: AuxStore, T>(
 
 	// the header is valid but let's check if there was something else already
 	// proposed at the same slot by the given author
-	if let Some(equivocation_proof) = check_equivocation(
+	let slot_number = pre_digest.slot_number();
+	let equivocation_proof = check_equivocation(
 		client,
 		slot_now,
-		pre_digest.slot_number(),
+		slot_number,
 		&header,
+		sig,
 		author,
-	).map_err(|e| e.to_string())? {
+	).map_err(|e| e.to_string())?;
+
+	if let Some(equivocation_proof) = equivocation_proof {
 		info!(
 			"Slot author {:?} is equivocating at slot {} with headers {:?} and {:?}",
 			author,
-			pre_digest.slot_number(),
-			equivocation_proof.fst_header().hash(),
-			equivocation_proof.snd_header().hash(),
+			slot_number,
+			equivocation_proof.first_header.hash(),
+			equivocation_proof.second_header.hash(),
 		);
+
+		// Submit a transaction reporting the equivocation.
+		let block_id = BlockId::number(client.info().best_number);
+		let maybe_report_transaction = client
+			.runtime_api()
+			.construct_equivocation_transaction_with_context(
```

Comment:
This is IMO not the right level of abstraction.

It's better to think of things in terms of making code testable: rather than a `SubmitExtrinsic` trait (which is itself better than passing a raw transaction queue), you should have a `SubmitEquivocationProof` trait, which deals with extrinsics internally.

This means that tests don't have to deal with extrinsics, which is way more nasty.